### PR TITLE
Rename to or from FileSystem

### LINKED ISSUE #88: `Path.rename(S3Path)` doesn't work
Crossing type boundaries to easily "upload" a file doesn't seem to work. Is there a better way to do this?

pathlib's path is just running their `_NormalAccessor` which is using `os.rename` it seems like. It doesn't really check the `target` in `Path.rename(target)`

### PR DESCRIPTION
**Pull request by petrus-v**
Proposal to fix #88 based on ~#106~ #111

### CODE REVIEW

**Comment by liormizr** [comment]
Hi @petrus-v 
This is a great Idea
Do you want to wait until we will finish talking about PR #106 ?
Or do you want to split them and look at this as a separate project?

**Comment by petrus-v** [comment]
Let's decide what we do on #106  first !

Fun fact this code is in production since this night ^^

Still getting some troubles to configure boto3 with scaleway with large files (24Go)

File: docs/advance.rst [hunk 1]
```diff
@@ -110,6 +110,80 @@ This example show how to specify default AWS S3 parameters, a `LocalStack`_ Buck
    >>> register_configuration_parameter(minio_bucket_path, resource=minio_resource)
 
 
+while using rename to/from S3 path we are using upload/download methods
+but s3 transfer configuration set in your `.aws/config` file won't be honored
+as this is not part of boto3_ but config interpretation is part of `aws-cli`_ in the
+`RuntimeConfig` class.
+
```
Comment (petrus-v) [inline docs/advance.rst:117]: I've to warn that rename from/to fs support depends on uri-pathlib-factory  library

File: s3path.py [hunk 2]
```diff
@@ -920,3 +986,28 @@ def is_symlink(self, *args, **kwargs):
...
+    if isinstance(target, str):
+        target = PathFactory(target)
+    if self.__class__ is S3Path and target.__class__ is not S3Path:
+        return self._rename_to_fs(target)
+    elif self.__class__ is not S3Path and target.__class__ is S3Path:
+        return target._rename_from_fs(self)
+    return original_path_rename(self, target)
+
+
+Path.rename = __mokey_patch_path_rename__
```
Comment (petrus-v) [inline s3path.py:1013]: I'll thinking to move this patch on uri-pathlib-factory library somehow

**Review by petrus-v — COMMENTED** [review]
I've rebased on #111 to test it but probably patching pathlib.Path.rename should be move in uri-pathlib-factory library as well.

Let's first finalize #111  before deepen on it here

File: s3path.py (hunk 2 above)
Comment (petrus-v) [reply s3path.py:1013]: What's could be great to achieved is something lire this (Assuming ftp backend is developed)

```python
Path("ftp://path/to/file.txt").rename("s3://buckekt/path/to/s3-file.txt")
```

In order to move files from any uri backend !

**Comment by liormizr** [comment]
Hi @petrus-v 
If you want to fix the issue when doing rename from s3 to local file system or the other way around that's nice
And that's the task that I thought you want to do.
But like I wrote to you in the URI lib factory PR I don't think that s3path don't need to require uri-lib-facroty
I think that it need to be the opposite 
Maybe this feature need to be in uri-lib-facroty if you want to support all rename options (move from s3 to ftp or filesystem or ....)

**Comment by petrus-v** [comment]
> If you want to fix the issue when doing rename from s3 to local file system or the other way around that's nice
> And that's the task that I thought you want to do.
> But like I wrote to you in the URI lib factory PR I don't think that s3path don't need to require uri-lib-facroty
> I think that it need to be the opposite
> Maybe this feature need to be in uri-lib-facroty if you want to support all rename options (move from s3 to ftp or filesystem or ....)

indeed this could solve the issue, thanks for feedback.

I haven't deep that much here, this works for the time being and don't expect to be merged as it anyway.

I've another branch that add copy method (ie: `my_path.copy(other_path)`) which is unfortunate not provided by pathlib interface and allows to copy file from s3 to filesystem that should be some how an equivalent use cases.

I'll think more to takes right decision on #111

**Pull request closed without merging**